M355 case lightimprovements

Bob-the-Kuhn opened the pull request:
updated code for M355 case light 

1. Made it so CASE_LIGHT_PIN can be defined either in config_adv or in
the pins_XXX.h file.  This was done by moving the enable function from
the pin definition to a flag.  That required a change to the
CONDITIONALS file.

2. Made it so that the case light responded to both the "P" and "S"
parameters no matter if the pins was digital only or if it had a PWM.
This was done by setting BOTH the ON/OFF flag and the brightness
variables when EITHER parameter is entered.

3. Made it so the brightness was also controlled by the invert flag.

4. Added print statements so that the brightness level was reported if a
PWM pin was being used.

Bob-the-Kuhn commented:
Splitting the M355 changes out of PR  #5668 

This PR is complete pending review.

manianac commented on Marlin/Marlin_main.cpp:
```diff
@@ -7598,15 +7598,27 @@ inline void gcode_M907() {
+    if (code_seen('S')) {
+      case_light_on = code_value_bool();
+      case_light_on ? case_light_brightness = 255 : case_light_brightness = 0;
+    }
```
Personally I'm against this.  If I specify P and S at the same time, S will always override P.  On top of that, why turn off the case light with brightness below 127?  I'll turn mine down to 64 or so to see the printer late at night without lighting up the whole room.

manianac reviewed (changes requested):
Drop suggested parameter changes

Bob-the-Kuhn commented:
If you're using a digital pin then P above 127 will turn it on, off otherwise.  The same thing as if a S1or a S0were issued.

If you're using a PWM then P does the standard brightness thing and S will set it at fully on or fully off.

I'd also scratched my head at why there were both S & P parameters.  I'm just trying to stay as true to the originator's intentions as I could (assuming I'm interpreting things properly).  Personally  I think that having just the S parameter is more like the M42 command.

manianac commented:
I agree, it should just be one 'S' parameter.

But say I set the brightness, and now specify S to turn it on and off.  You've overwritten any brightness setting on line 7613.  There's also no need for the >127 check, as analogWrite does that for us.

Arduino source:
```c
		        case NOT_ON_TIMER:
			default:
				if (val < 128) {
					digitalWrite(pin, LOW);
				} else {
					digitalWrite(pin, HIGH);
				}
```

I figure we either move to an 'S' parameter, or just change the beginning back to 
```c
    if (code_seen('P')) case_light_brightness = code_value_byte();
    if (code_seen('S')) case_light_on = code_value_bool();
```

Bob-the-Kuhn commented:
Another option is to disallow having both parameters on the same command.

My preference is to just use the S parameter only.

The original code just sent ON or OFF to the serial port.  I'd rather keep sending the brightness to the serial port when the pin is a PWM.

Bob-the-Kuhn commented:
I've changed everything over to using just the **S** parameter.  I've also consolidated all the init stuff into one area and removed the **case_light_on** flag from Marlin_main.cpp.  

The **case_light_on** flag is still in the LCD section.  I tried to change the LCD code as little as possible.

Please take a close look at the LCD changes.  I have no way to test them.

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -614,11 +614,13 @@ void kill_screen(const char* lcd_msg) {
 
-    extern bool case_light_on;
+    extern uint8_t case_light_brightness;  
+    bool case_light_on (case_light_brightness >= 128) ? true : false;
```
syntax error - missing '='

thinkyhead commented:
The two concepts —on/off state (`P`) and brightness (`S`)— are not just because of the digital versus analog pins. They aren't meant to be analogous to each other but complimentary. The only thing missing is that `M355` doesn't do a check for a PWM-capable pin to leave out the call to `analogWrite` when not needed.

I still like the idea of having both concepts — of brightness and on/off state. But some enhancements make sense, so you can use `P` optionally:

- `M355 S0` should be like `M355 P0`
- `M355 S1` and higher should turn on the light (`M355 P1`), if not on

shacal commented:
Well, definitely see only one S param.
I think we have couple cases.

PWM pin and user sends on/off.
Digital pin and user sends on/off.
PWM pin and user sends 0-100% brightness.
Digital pin and user sends 0-100% brightness.

Why don't we just have S value with range of 0-100.
If pin is digital (we can ask that from arduino framework), use digitalWrite(s > 0 ? HIGH : LOW)
If pin is analog  (we can ask that from arduino framework), use analogWrite( 0 - 100% )...

This is how many home automation lights work, as an example.

Or did I miss someting? 🐳

manianac commented:
With the exception of flow and feed rate, I think all byte adjustables have
been 0-255. See any other pwm pins (fans come to mind)

On Sat, Jan 14, 2017, 9:23 AM Remo Kallio <notifications@github.com> wrote:

> Well, definitely see only one S param.
> I think we have couple cases.
>
> PWM pin and user sends on/off.
> Digital pin and user sends on/off.
> PWM pin and user sends 0-100% brightness.
> Digital pin and user sends 0-100% brightness.
>
> Why don't we just have S value with range of 0-100.
> If pin is digital (we can ask that from arduino framework), use
> digitalWrite(s > 0 ? HIGH : LOW)
> If pin is analog (we can ask that from arduino framework), use
> analogWrite( 0 - 100% )...
>
> This is how many home automation lights work, as an example.
>
> Or did I miss someting? 🐳
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/5685#issuecomment-272627200>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB_uHRauVqjU0kzI6THbZgdwz5Ne1scQks5rSNpfgaJpZM4LiDF8>
> .
>

Bob-the-Kuhn commented:
I'd really like to keep the interface consistent with other Marlin commands.

Right now M355 acts just like M42 with but with additional reporting.

If we go away from the M42 format then
1. Is there another Marlin command format that we should emulate?
2. Should we also change M42 to use the new format?

Bob-the-Kuhn commented:
I see where the originator got the S0 off, S1 on and P for brightness.  The [RepRap Wiki](http://reprap.org/wiki/G-code) has the S0, S1 which meant that he needed a different parameter for brightness.

It's easy to revert the code back to the P for brightness method.

My vote would be to stay with the current S for brightness. method.

dot-bob commented:
Playing off the idea of thinkyhead's comment.  The led driver in the Linux kernel treats digital and PWM pins in this manner:

Value of 0:
PWM = off
Digital = off

Value X where (X > 0) && (X <= 255)
PWM  = Brighness of (X/255 * 100)%
Digital = full brightness (digital pin stays asserted)

So I would propose this:

For M355 S1 is equivlent to M355 P255
M355 S0 -> M355 P0

And if they have a digital pin and use M355 P1 through P255
use the Linux kernel method above and the digital pin stays asserted.

If they specify both S and P prams in the same command then P overrides.

And the interface stays consistent with other commands.

Bob-the-Kuhn commented:
That's easy to implement.  Just revert & move the "P" check to after the "S" check (and correct the comment).

The only question I have is the value of P to turn a digital pin on & off.
- M42 system:  If P is 128 & above the pin is set to ON.  If P is 127 or less then the pin is set to OFF. 
- Linux system: If P is zero then the light is OFF, P  of 1 or above turns the light ON.

Is there somewhere in Marlin that the user is expecting to use the 1 decision point?   If not then we should stick with the M42 system (and not introduce another system).

FYI - I just looked at the analogWrite routine.  If the pin isn't a PWM then it does a digitalWrite of 1 if the value is 128 & above, 0 otherwise.  Turning the light ON if P is 1 or above requires just a couple of lines of code to implement (basically always do a digitalWrite if the pin is digital, always do an analogWrite if the pin is PWM.

Bob-the-Kuhn commented:
My preferences are :
1. M42 system - one parameter (S) & on/off for digital at 127/128
2. Two parameters - S is Boolean, P overrides & decision is at 127/128
3. Two parameters - S is Boolean, P overrides & decision is at 0/1

thinkyhead committed:
updated code for M355 case light

1. Made it so CASE_LIGHT_PIN can be defined either in config_adv or in
the pins_XXX.h file.  This was done by moving the enable function from
the pin definition to a flag.  That required a change to the
CONDITIONALS file.

2. Made it so that the case light responded to both the "P" and "S"
parameters no matter if the pins was digital only or if it had a PWM.
This was done by setting BOTH the ON/OFF flag and the brightness
variables when EITHER parameter is entered.

3. Made it so the brightness was also controlled by the invert flag.

4. Added print statements so that the brightness level was reported if a
PWM pin was being used.

==================================================

changed to "S" only

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -8281,29 +8271,33 @@ inline void gcode_M907() {
+    if (code_seen('S')) {
+      case_light_brightness = code_value_byte();
+      update_case_light();
+      if (NOT_ON_TIMER == digitalPinToTimer(CASE_LIGHT_PIN)) {
```
Wow, that `NOT_ON_TIMER` trick is pretty handy. Is that how we can test any pin for PWM-capability? And, do we need to filter out the timers that are used by Marlin?

Bob-the-Kuhn replied:
I had to refresh my memory.  It's just `#define NOT_ON_TIMER 0` .

I doubt that the average user would know enough about PWMs and how Marlin uses the counters to be able to reliably pick a usable hardware PWM.  It's probably a good idea to create a useable_hardware_PWM(uint8_t pin) routine.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -8281,29 +8271,33 @@ inline void gcode_M907() {
+        SERIAL_ECHOPGM("Case lights ");
+        (case_light_brightness >= 128) ? SERIAL_ECHOLNPGM("on") : SERIAL_ECHOLNPGM("off");
+      }
+      else SERIAL_PROTOCOLLNPAIR("echo:Case lights ", case_light_brightness);
```
For consistency I think it's best to use…
```cpp
else {
  SERIAL_ECHO_START;
  SERIAL_PROTOCOLLNPAIR("Case lights ", case_light_brightness);
}
```

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -8281,29 +8271,33 @@ inline void gcode_M907() {
 #if HAS_CASE_LIGHT
 
-  uint8_t case_light_brightness = 255;
+  uint8_t case_light_brightness;
```
If this were meant to determine the initialized value of `case_light_on` then it would be initialized here to something based on `CASE_LIGHT_DEFAULT_ON`. But, since `case_light_brightness` is being initialized in `setup()` and then `update_case_light` is being called, this is fine to leave as-is.

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -684,11 +684,13 @@ void kill_screen(const char* lcd_msg) {
 
-    extern bool case_light_on;
+    extern uint8_t case_light_brightness;  
+    bool case_light_on = (case_light_brightness >= 128) ? true : false;
```
Since `setup` initializes `case_light_on` by calling `update_case_light` (and since at this point `case_light_brightness` is always `0`) this should be left uninitialized here.
```cpp
bool case_light_on;
```
Note also, that if a ternary ends with `true : false` you can drop the ternary and just use:
```cpp
 bool case_light_on = case_light_brightness >= 128;
```

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -684,11 +684,13 @@ void kill_screen(const char* lcd_msg) {
     extern void update_case_light();
 
     void toggle_case_light() {
       case_light_on = !case_light_on;
```
You can drop this line if you have `update_case_light` handle updating the state of `case_light_on` based on `case_light_brightness`.

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -684,11 +684,13 @@ void kill_screen(const char* lcd_msg) {
 
     void toggle_case_light() {
       case_light_on = !case_light_on;
+      case_light_brightness -= 255;
```
Perhaps we should leave `case_light_brightness` as it is here. If the user "toggles" the light, it would be nice if it came back on with the same brightness level it had when it was turned off.

And wouldn't this be…?
```cpp
case_light_brightness = 255 - case_light_brightness;
```
…or better…
```cpp
case_light_brightness ^= 0xFF;
```

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -11133,6 +11127,10 @@ void setup() {
+    case_light_brightness = 0;   // on power up or reset, default to off
+    #if ENABLED(CASE_LIGHT_DEFAULT_ON)
+      case_light_brightness = 255;
+    #endif
```
Generally, to "reduce redundancy" the style has been to break a single code line:
```cpp
case_light_brightness =
  #if ENABLED(CASE_LIGHT_DEFAULT_ON)
    255
  #else
    0
  #endif
;
```
I wish we could use `ENABLED` inside code expressions. It would simplify things.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -8281,29 +8271,33 @@ inline void gcode_M907() {
-    digitalWrite(CASE_LIGHT_PIN, case_light_on != INVERT_CASE_LIGHT ? HIGH : LOW);
-    analogWrite(CASE_LIGHT_PIN, case_light_on != INVERT_CASE_LIGHT ? case_light_brightness : 0);
+    digitalWrite(CASE_LIGHT_PIN, (case_light_brightness >= 128) != INVERT_CASE_LIGHT ? HIGH : LOW);
+    analogWrite(CASE_LIGHT_PIN, INVERT_CASE_LIGHT ? 255 - case_light_brightness : case_light_brightness );
```
I think the light toggle should only change `case_light_on`, and then this code should be:
```cpp
const uint8_t clb = case_light_on ? case_light_brightness : (INVERT_CASE_LIGHT ? 255 : 0);
digitalWrite(CASE_LIGHT_PIN, (clb >= 128) != INVERT_CASE_LIGHT ? HIGH : LOW);
analogWrite(CASE_LIGHT_PIN, INVERT_CASE_LIGHT ? 255 - clb : clb);
```

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -8281,29 +8271,33 @@ inline void gcode_M907() {
+      if (NOT_ON_TIMER == digitalPinToTimer(CASE_LIGHT_PIN)) {
+        SERIAL_ECHO_START;
+        SERIAL_ECHOPGM("Case lights ");
+        (case_light_brightness >= 128) ? SERIAL_ECHOLNPGM("on") : SERIAL_ECHOLNPGM("off");
```
When the pin for the case light is non-PWM then it makes sense that the brightness set by `S` affect `case_light_on` directly. So…
```cpp
case_light_on = case_light_brightness >= 128;
serialprintPGM(case_light_on ? PSTR("on") : PSTR("off"));
```

thinkyhead commented:
> My preferences are…

If we were going to go for a single parameter, then `S0` should turn the light off, `S1` to `S255` should turn on a non-PWM light or set the intensity of a PWM light.

If going for two parameters, I prefer `S` as a switch to turn the light on and off (preserving the brightness setting), while `P` is used to set the brightness.
- If the `P` parameter is given by itself, then `P0` should set `case_light_on` to `false` in addition to setting `case_light_brightness` to zero.
- If the pin is non-PWM then `P1` to `P255` should just turn the light on to its full intensity (because digital).
- If the `S` parameter is given while `case_light_brightness` is zero (or very low) then it should set `case_light_brightness` to `128` so it's clear that the command did work and was not simply ignored.
- `M355 S1 P0` should set `case_light_brightness` to 0 and `case_light_on` to `false`.

The single parameter version is simpler, so I tend to prefer that!

Bob-the-Kuhn commented:
Single S parameter it is.

Turn non-PWM on at 1 vs. 128 - personally I don't care which one it is.  From a user standpoint I think we should have M42 and M355 behave the same.

Bob-the-Kuhn commented:
Here's another idea.  Let's get rid of the toggle function entirely.  It's making things complicated & I don't think it's needed.

To that end I made the following changes:
- in ultralcd.cpp replaced the toggle LED menu item with a brightness menu item
- in marlin_main.cpp got rid of the case_light_on variable.

The on/off feature when using a non-PWM pin should still be there.

I've done a little testing with a PWM pin and an LCD & it appears to work.

The code is rough.  I wanted to get feedback on the concept before moving on to work on Scott's priorities for 1.1.0.

I need to clean up the code and probably move the LCD brightness item into a submenu.  We can review it in May/June when things calm down

----

I've implemented all of Scott's suggestions except for the toggle function.

Creating a useable_hardware_PWM() function was harder than I expected.  I ended up modifying some code from pinDebugging.  I also left my cheat sheet on usable PWMs in the comments.

I also created a define for the zero/one decision threshold so we could easily change it.

----

It looks like Scotts changes from this morning have disappeared.  

When I went to synch this commit I got back what looked like a second commit with 100+ files in error.  Some of these changes could not be discarded so I did a `git reset --hard head` followed by a rebase and then a `git push -f`.

As best I can tell I had the local branch open when Scott did his commit this morning.  That meant I didn't have a local copy of Scott's commit.  The  `git reset --hard head` basically threw away everything that my local branch didn't know about.

Roxy-3D commented:
> git push -f

So....   I don't pretend to know much about Git.   But the -f option isn't something to be used lightly.   The 'f' is for 'Force'.   That is always dangerous (like on an rm -f command....)

Git is not intuitive!

GitHub Desktop needs a couple of buttons called "Do the right thing"   and "Resolve all Conflicts"

Bob-the-Kuhn commented:
Didn't have much of a choice.

The only other option I had was to ask Scott to fix it.  I'd rather use my time to fix any unintended consequences than pull Scott away from the high priority stuff.

I'll gladly re-create Scott's commit if that's what he wants.

Roxy-3D commented:
How would you re-create Scott's commit?   Are you thinking about doing a fork of the current code base and then manually reconstructing the changes?  And then use that to create a new pull request?

The reason I'm asking is I think the -f makes everything further down the branch 'unavailable'.   I'm not sure I understand this option correctly.   That is why I'm asking.

Bob-the-Kuhn commented:
I'd manually recreate it and then just upload it to this PR.

thinkyhead commented:
> So.... I don't pretend to know much about Git. But the -f option isn't something to be used lightly. 

I use it several dozen times a day, whenever commits are removed or squashed (combined). It's perfectly safe, and necessary if you like having a nice neat commit log rather than a bunch of little commits that are just fixes of typos.

> Git is not intuitive!

Correct. It requires study and practice.

> I think the -f makes everything _further down the branch_ 'unavailable'

It doesn't do anything of the sort. "Where knowledge is absent, superstition prevails."

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -8556,29 +8545,137 @@ inline void gcode_M907() {
+   */
+
+
+    switch(digitalPinToTimer(pin)) {
```
Add a space between `switch` and the opening parenthesis.
Only _functions_ should snuggle with the opening parenthesis.
http://marlinfw.org/docs/development/coding_standards.html#spacing

Bob-the-Kuhn commented:
fixed the formatting, rebased & squashed

Bob-the-Kuhn committed:
fix copy & paste mistake

Bob-the-Kuhn commented:
This PR has been replaced by PR #6797
